lib/hathifiles_database/db/connection.rb

```diff
@@ -41,35 +44,43 @@ def initialize(connection_string, logger: LOGGER)
       # Update the tables from a file just by directly deleting/inserting
       # the values. It's slow, but not so slow that it's not fine for a normal
       # nightly changefile, and it's a lot less screwing around.
-      def update_from_file(filepath, linespec = LineSpec.default_linespec, logger: Constants::LOGGER)
+      def update_from_file(filepath, linespec = LineSpec.default_linespec, logger: Constants::LOGGER,
+        delta: HathifilesDatabase::Delta.new)
         # path = Pathname.new(filepath)
         datafile = HathifilesDatabase::Datafile.new(filepath, linespec, logger: logger)
-        upsert(datafile)
+        upsert(datafile, delta: delta)
       end
 
       # Update the database with data from a bunch of HathifileDatabase::Line
       # objects.
       # @param [Enumerable<HathifileDatabase::Line>] lines An enumeration of
       #   lines (generally just a datafile, which has the right interface)
-      def upsert(lines)
-        slice_size = 100
-        log_report_chunk_size = 5000
+      # @param [HathifilesDatabase::Delta] changed and deleted HTIDs
+      def upsert(lines, delta:)
+        lines_seen = 0
+        rows_added = 0
         mysql_set_foreign_key_checks(:on)
         @rawdb.transaction do
-          records = 0
-          lines.each_slice(slice_size) do |lns|
-            delete_existing_data(lns)
-            add(lns)
-            records += 1
-            logger.info "Inserted/replaced #{records * slice_size} records" if (records * slice_size) % log_report_chunk_size == 0
+          lines.each_slice(UPSERT_SLICE_SIZE) do |slice|
+            lines_seen += slice.size
+            # Select only the records that have changed.
+            slice.select! { |line| delta.updated?(line.htid) }
+            delete_existing_data(slice)
+            add(slice)
+            rows_added += slice.size
+            if lines_seen % LOG_REPORT_CHUNK_SIZE == 0
+              logger.info "(upsert) records inserted/replaced: #{rows_added}/#{lines_seen}"
+            end
           rescue HathifilesDatabase::Exception::WrongNumberOfColumns => e
             logger.error e
           rescue Sequel::DatabaseError => e
             logger.error e
             abort
           end
-          logger.info "Total inserted: #{records * slice_size}"
         end
+        logger.info "(upsert final) records inserted/replaced: #{rows_added}/#{lines_seen}"
```

Comment:
No need to do this now, but we definitely want to get prometheus/pushgateway stuff on this (see #8)